R/combine-vcf-with-pgs.R

```diff
@@ -69,19 +82,97 @@ combine.vcf.with.pgs <- function(vcf.data, pgs.weight.data) {
+        # rename ID column to match pre-merge pgs weight data
+        missing.snp.merged.data$ID <- missing.snp.merged.data$ID.pgs;
+        missing.snp.merged.data$ID.vcf <- NULL;
+        missing.snp.merged.data$ID.pgs <- NULL;
+        # extract only original pgs weight columns from merged missing SNP data (for neater second merge)
+        missing.snp.pgs.weight.data <- subset(missing.snp.merged.data, select = colnames(pgs.weight.data));
+        rm(missing.snp.merged.data);
+
+        # Split VCF$ID column into separate rows for each rsID (multiple rsIDs are separated by ;)
+        # most efficient way to do this is to use the data.table package
+        if (any(grepl(';', vcf.data$ID))) {
+            data.table::setDT(vcf.data);
+            split.rsid.vcf.data <- merge(
+                x = vcf.data,
+                # split only entries with multiple rsIDs, save in new column, and merge back with the original data
+                y = vcf.data[grepl(';', ID), strsplit(as.character(ID), ';', fixed = TRUE), by = .(Indiv, CHROM, POS)
+                    ][,.(new.ID = V1, Indiv, CHROM, POS)],
+                by = c('CHROM', 'POS', 'Indiv'),
+                all = TRUE
+                );
+            # replace entries with multiple rsIDs with the new, split, rsID
+            split.rsid.vcf.data <- split.rsid.vcf.data[!is.na(new.ID), ID := new.ID][, new.ID := NULL];
+            } else {
+            split.rsid.vcf.data <- vcf.data;
+            }
+
+        # merge missing SNP data on split rsID
+        merged.vcf.with.missing.pgs.data <- merge(
+            x = missing.snp.pgs.weight.data,
+            y = split.rsid.vcf.data,
+            suffixes = c('.pgs', '.vcf'),
+            by = 'ID',
+            all.x = TRUE
+            );
+        rm(split.rsid.vcf.data);
+
+        # update missing SNP index
+        second.merge.missing.pgs.snp.index <- is.na(merged.vcf.with.missing.pgs.data$REF);
+
+        if (all(missing.pgs.snp.index) & (sum(missing.pgs.snp.index) == sum(second.merge.missing.pgs.snp.index))) {
+            stop('All PGS SNPs are missing from the VCF, terminating merge.');
+            } else if (any(second.merge.missing.pgs.snp.index)) {
+                warning(paste('PGS is missing', sum(second.merge.missing.pgs.snp.index)), ' SNPs from VCF');
+                missing.snp.data <- merged.vcf.with.missing.pgs.data[second.merge.missing.pgs.snp.index, ];
+            } else {
+            missing.snp.data <- NULL;
+            }
+
+        # keep coordinates from VCF data for matched SNPs with coordinate mismatch
+        merged.vcf.with.missing.pgs.data[!is.na(merged.vcf.with.missing.pgs.data$REF), 'CHROM'] <- merged.vcf.with.missing.pgs.data[!is.na(merged.vcf.with.missing.pgs.data$REF), 'CHROM.vcf'];
+        merged.vcf.with.missing.pgs.data[!is.na(merged.vcf.with.missing.pgs.data$REF), 'POS'] <- merged.vcf.with.missing.pgs.data[!is.na(merged.vcf.with.missing.pgs.data$REF), 'POS.vcf'];
```

Comment:
I wonder if this can be done all at once, without calling `is.na(merged.vcf.with.missing.pgs.data$REF)` multiple times.